### PR TITLE
IRMA Server Flake

### PR DESCRIPTION
**Pull request by raghuramlavan**
Request for review. The IRMA server is GO lang based and requires postgresql and MailHog only for testing.

### CODE REVIEW

**Comment by cdepillabout** [comment]
This PR is only packaging the derivation for the server part of IRMA.  @raghuramlavan will send a follow-up PR that contains a NixOS module for running the server.

File: flake.nix [hunk 1]
```diff
@@ -30,21 +26,23 @@
...
+          src = irma-server-src;
 
-          buildInputs = [ autoconf automake gettext gnulib perl gperf texinfo help2man ];
+          vendorSha256 =  "sha256-9JZKl5hm3qtbYNwPFyDSzndhw6/Qr2DN/CY3eSNDMxU=";
 
-          preConfigure = ''
-            mkdir -p .git # force BUILD_FROM_GIT
-            ./bootstrap --gnulib-srcdir=${gnulib-src} --no-git --skip-po
+         checkPhase = ''
+          echo "Testing requres postgresql and MailHog which are not required for operation hence it is done only in nix develop"
```
Comment (DavHau) [inline flake.nix]: Better to use `doCheck = false` and put this text as a comment above it.

File: flake.nix [hunk 2]
```diff
@@ -53,67 +51,112 @@
...
+        });
 
       # Tests run by 'nix flake check' and by Hydra.
       checks = forAllSystems (system: {
-        inherit (self.packages.${system}) hello;
+         inherit (self.packages.${system}) irma-server;
+         test =
+         with nixpkgsFor.${system};
+        stdenv.mkDerivation {
+            name = "irma-server-test-${version}";
```
Comment (DavHau) [inline flake.nix]: indentation seems off here

File: shell.nix [hunk 3]
```diff
@@ -1,3 +1,39 @@
-(import (fetchTarball https://github.com/edolstra/flake-compat/archive/master.tar.gz) {
-  src = builtins.fetchGit ./.;
-}).shellNix
+{ pkgs ? import <nixpkgs> {} }:
```
Comment (DavHau) [inline shell.nix]: Seems redundant. You are already using flakes with devShell. This file should just reference the existing devShell via flake-compat like in the original project-template.

File: flake.nix [hunk 4]
```diff
@@ -30,21 +26,23 @@
...
           '';
 
           meta = {
-            homepage = "https://www.gnu.org/software/hello/";
-            description = "A program to show a familiar, friendly greeting";
+            homepage = "https://privacybydesign.foundation/irma-explanation/";
+            description = "This program is IRMA server.
+            IRMA stands for I Reveal My Attributes. 
+            IRMA empowers you to disclose online,
+            via your mobile phone, certain attributes of yourself";
```
Comment (DavHau) [inline flake.nix]: Better to use multi line string with double single quotes.

**Review by DavHau — COMMENTED** [review]
I left some comments.
As it is a single package and your test only involves that package the testing should probably be done in checkPhase of the original package's derivation. I think there is no reason to use an additional flake check at all.

**Comment by DavHau** [comment]
`nix run .` fails with:
```
error: unable to execute '/nix/store/lwhkbcp0k5y3dagml154h3xkf4a84cwf-irma-server-20210727/bin/irma-server': No such file or directory
```
This can be fixed by defining a defaultApp.

File: flake.nix [hunk 5]
```diff
@@ -53,67 +51,112 @@
...
+          
+            vendorSha256 =  "sha256-9JZKl5hm3qtbYNwPFyDSzndhw6/Qr2DN/CY3eSNDMxU=";
 
             unpackPhase = "true";
 
-            buildPhase = ''
-              echo 'running some integration tests'
-              [[ $(hello) = 'Hello, world!' ]]
+            postInstall = ''
+              export PGHOST=$HOME/postgres
```
Comment (happysalada) [inline flake.nix:127]: How about trying to run those as vm test?
Grafana has a good example of a test with a service that has a db
https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/grafana.nix
Running all those stateful changes in a bash script feels like it depends on implementation details. Implementing it in with nixos test should be less brittle when upgrades happen.

File: shell.nix [hunk 6]
```diff
@@ -1,3 +1,39 @@
-(import (fetchTarball https://github.com/edolstra/flake-compat/archive/master.tar.gz) {
```
Comment (happysalada) [inline shell.nix]: can we keep the original shell.nix and default.nix, those are there for compatibility reasons with non flake enabled systems.

File: shell.nix [hunk 7]
```diff
@@ -1,3 +1,39 @@
...
+       repo = "irmago";
+       rev = "0c153a97d013162688e77f104c65bd246b5d5323";
+       sha256 = "sha256-Y+DUFOlgmyHloQwq0MTs/oabLf+8NHVElguoM9EVvmo=";
+     };
+
+     shellHook = ''
+       export PGHOST=./postgres
+       export PGDATA=$PGHOST/data
+       export PGDATABASE=postgres
+       export PGLOG=$PGHOST/postgres.log
```
Comment (happysalada) [inline shell.nix]: Setting the environment variables in the shellHook is good.
Is the PGLOG var only needed for postgres ?
If it is you don't need to set another environment variable and can directly replace the value inside the pg_ctl call

However all the mutable changes below should be part of the readme, not part of the hook.
The way you have it setup here, those changes run on every time you cd in the directory (if direnv is installed for example). There could be ways to mitigate this, but setups steps should most likely be part of the readme.

Comment (DavHau) [reply shell.nix]: You could make the whole setup procedure a bash function and then the user can just execute `init-postgres`, or something similar. Then it is not executed on every session.

Comment (raghuramlavan) [reply shell.nix]: The initialisation of postgreSql and MailHog is only for testing?. If we remove the test componet in the flake then we can seperate it. I was of the understanding that there is a requirement for testing in an fully automated manner..

Comment (happysalada) [reply shell.nix]: Postgres is needed for dev environment as well right ?
a bash function sounds good to me.

File: flake.nix (hunk 1 above)
Comment (raghuramlavan) [reply flake.nix]: removed.

File: flake.nix (hunk 2 above)
Comment (raghuramlavan) [reply flake.nix]: fixed

File: shell.nix (hunk 3 above)
Comment (raghuramlavan) [reply shell.nix]: the default file is restored,

File: flake.nix (hunk 4 above)
Comment (raghuramlavan) [reply flake.nix]: fixed as suggessted

File: shell.nix (hunk 6 above)
Comment (raghuramlavan) [reply shell.nix]: default files restored

**Comment by raghuramlavan** [comment]
> `nix run .` fails with:
> 
> ```
> error: unable to execute '/nix/store/lwhkbcp0k5y3dagml154h3xkf4a84cwf-irma-server-20210727/bin/irma-server': No such file or directory
> ```
> 
> This can be fixed by defining a defaultApp.

i have added the code for default app

File: flake.nix (hunk 5 above)
Comment (raghuramlavan) [reply flake.nix:127]: will try to implement the vm test

**Comment by raghuramlavan** [comment]
@cdepillabout the remarks from the review have been fixed and NixOS module for the server is done. we can add ready for review lable.

**Comment by cdepillabout** [comment]
@raghuramlavan Can you merge in this PR and send a quick message on ngi-nix/ngi#42?

**Pull request merged**